### ISSUE TITLE
Unable to import a module in VS code when using conda

### ISSUE THREAD

**Issue by dylan-behr** [microsoft/vscode-jupyter#6306]
## Environment data

-   VS Code version:  1.57
-   Jupyter Extension version (available under the Extensions sidebar): v2021.6.999406279
-   Python Extension version (available under the Extensions sidebar): v2021.5.926500501
-   OS (Windows | Mac | Linux distro) and version: Windows 10
-   Python and/or Anaconda version: Python 3.7.10, conda 4.10.1, numpy 1.19.1
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): conda
-   Jupyter server running: Local 

## Expected behaviour

import numpy successfully

## Actual behaviour

ImportError: DLL load failed: The specified module could not be found.

## Steps to reproduce:

1. Open Jupyter notebook in VSCode with conda env activated

2. Attempt to import numpy

## Logs

<details>

<summary>

C:\Users\dylan\Anaconda3\lib\site-packages\numpy\__init__.py:138: UserWarning: mkl-service package failed to import, therefore Intel(R) MKL initialization ensuring its correct out-of-the box operation under condition when Gnu OpenMP had already been loaded by Python process is not assured. Please install mkl-service package, see http://github.com/IntelPython/mkl-service
  from . import _distributor_init
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
~\Anaconda3\lib\site-packages\numpy\core\__init__.py in <module>
     21 try:
---> 22     from . import multiarray
     23 except ImportError as exc:

~\Anaconda3\lib\site-packages\numpy\core\multiarray.py in <module>
     11 
---> 12 from . import overrides
     13 from . import _multiarray_umath

~\Anaconda3\lib\site-packages\numpy\core\overrides.py in <module>
      6 
----> 7 from numpy.core._multiarray_umath import (
      8     add_docstring, implement_array_function, _get_implementing_args)

ImportError: DLL load failed: The specified module could not be found.

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-d4cdadb62aa7> in <module>
----> 1 import numpy

~\Anaconda3\lib\site-packages\numpy\__init__.py in <module>
    138     from . import _distributor_init
    139 
--> 140     from . import core
    141     from .core import *
    142     from . import compat

~\Anaconda3\lib\site-packages\numpy\core\__init__.py in <module>
     46 """ % (sys.version_info[0], sys.version_info[1], sys.executable,
     47         __version__, exc)
---> 48     raise ImportError(msg)
     49 finally:
     50     for envkey in env_added:

ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.7 from "C:\Users\dylan\Anaconda3\python.exe"
  * The NumPy version is: "1.19.1"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: DLL load failed: The specified module could not be found.

(<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Jupyter</code>)
</summary>

Info 2021-06-16 12:21:13: Executing cell 167426b3-8f06-4b6a-95ce-7a22e961e348
Info 2021-06-16 12:21:13: Submitting code for 3ce6dc9a-8ea8-42cf-9aa9-bcae4c4ffb1a
Info 2021-06-16 12:21:13: UpdateWorkingDirectoryAndPath
Info 2021-06-16 12:21:13: changeDirectoryIfPossible
Info 2021-06-16 12:21:13: Kernel switching to busy
Info 2021-06-16 12:21:13: Kernel switching to idle
Info 2021-06-16 12:21:13: Preferred kernel for untitled:/c%3A/Users/dylan/Documents/Python_scripts/UCL/Untitled-2.ipynb is undefined
Info 2021-06-16 12:21:13: Kernel switching to busy
Info 2021-06-16 12:21:13: Preferred kernel for untitled:/c%3A/Users/dylan/Documents/Python_scripts/UCL/Untitled-2.ipynb is undefined
Info 2021-06-16 12:21:13: Preferred kernel for untitled:/c%3A/Users/dylan/Documents/Python_scripts/UCL/Untitled-2.ipynb is undefined
Info 2021-06-16 12:21:14: Preferred kernel for untitled:/c%3A/Users/dylan/Documents/Python_scripts/UCL/Untitled-2.ipynb is undefined
Info 2021-06-16 12:21:14: Finished execution for 167426b3-8f06-4b6a-95ce-7a22e961e348
Info 2021-06-16 12:21:14: Finished executing cell 167426b3-8f06-4b6a-95ce-7a22e961e348
Info 2021-06-16 12:21:14: Kernel switching to idle

**Comment by rchiodo** [comment]
Thanks for the bug. This doesn't seem to be a problem with the jupyter extension, but rather your python environment.

This SO issue sounds like what your problem is:
https://stackoverflow.com/questions/36778066/importerror-dll-load-failed-when-importing-numpy-installed-in-conda-virtual-env

**Comment by dylan-behr** [comment]
@rchiodo Thanks for looking at the issue.

I believe it is a jupyter extension issue as the numpy import works fine in an ipython terminal.
The suggestions in the link do not help me resolve the issue.

**Comment by rchiodo** [comment]
That might happen if the kernel we're running is not the same environment as where you're running IPython.

Can you verify it's the same environment? For example, run 
```python
import sys
print(sys.executable)
```
in both to see if they match.

If those match, then check ```sys.path``` and ```os.environ```

**Comment by dylan-behr** [comment]
Hi Rich,

For whatever reason the import is now successful.. I'm not sure which step in particular did this.
Nonetheless, thanks very much for your help!

They were indeed the same executable running in IPython and Jupyter.

Best wishes,
Dylan

________________________________

**Comment by dylan-behr** [comment]
Dear Rich,

I've encountered the same error again but with matplotlib i.e. ImportError: DLL load failed.

In VSCode Jupyter extension I am using the same executable as in my IPython terminal (sys.executable returns 'C:\\Users\\dylan\\Anaconda3\\python.exe')
 However sys.path returns differently:
VSCode:
'c:\\Users\\dylan\\Documents\\Python_scripts\\UCL',
 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles',
 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles\\lib\\python',
 'C:\\Users\\dylan\\Anaconda3\\python37.zip',
 'C:\\Users\\dylan\\Anaconda3\\DLLs',
 'C:\\Users\\dylan\\Anaconda3\\lib',
 'C:\\Users\\dylan\\Anaconda3',
 '',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\extensions',
 'C:\\Users\\dylan\\.ipython'

IPython:
'C:\\Users\\dylan\\Anaconda3\\Scripts',
 'C:\\Users\\dylan\\Anaconda3\\python37.zip',
 'C:\\Users\\dylan\\Anaconda3\\DLLs',
 'C:\\Users\\dylan\\Anaconda3\\lib',
 'C:\\Users\\dylan\\Anaconda3',
 '',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\extensions',
 'C:\\Users\\dylan\\.ipython'


Should I add the missing 'C:\\Users\\dylan\\Anaconda3\\Scripts' to my VSCode path?

os.environ returns an item
'PYTHONPATH': 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles;c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles\\lib\\python'
amongst other items - is this also a problem?

Best,
Dylan
________________________________

**Comment by rchiodo** [comment]
PYTHONPATH should have those items added. That's how we load internal python modules.

The missing syspath thing is weird though. Not sure how this one isn't showing up:
```
C:\\Users\\dylan\\Anaconda3\\Scripts
```

That might be conda init? Conda init has to be run with the shell we're using to start the kernel with. It might be you haven't run it yet? And so when we try to get the conda environment something is missing.

Can you try running
```
conda init cmd.exe
conda init powershell
```

From an anaconda prompt, then try running VS code again and see if the sys path is fixed?

If that doesn't work you can add that scripts path manually.

**Comment by dylan-behr** [comment]
Hi Rich,

I tried running the commands you suggested and they returned the following:

(base) C:\Users\dylan>conda init cmd.exe
no change     C:\Users\dylan\Anaconda3\Scripts\conda.exe
no change     C:\Users\dylan\Anaconda3\Scripts\conda-env.exe
no change     C:\Users\dylan\Anaconda3\Scripts\conda-script.py
no change     C:\Users\dylan\Anaconda3\Scripts\conda-env-script.py
no change     C:\Users\dylan\Anaconda3\condabin\conda.bat
no change     C:\Users\dylan\Anaconda3\Library\bin\conda.bat
no change     C:\Users\dylan\Anaconda3\condabin\_conda_activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\rename_tmp.bat
no change     C:\Users\dylan\Anaconda3\condabin\conda_auto_activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\conda_hook.bat
no change     C:\Users\dylan\Anaconda3\Scripts\activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\deactivate.bat
no change     C:\Users\dylan\Anaconda3\Scripts\activate
no change     C:\Users\dylan\Anaconda3\Scripts\deactivate
no change     C:\Users\dylan\Anaconda3\etc\profile.d\conda.sh
no change     C:\Users\dylan\Anaconda3\etc\fish\conf.d\conda.fish
no change     C:\Users\dylan\Anaconda3\shell\condabin\Conda.psm1
no change     C:\Users\dylan\Anaconda3\shell\condabin\conda-hook.ps1
no change     C:\Users\dylan\Anaconda3\Lib\site-packages\xontrib\conda.xsh
no change     C:\Users\dylan\Anaconda3\etc\profile.d\conda.csh
modified      HKEY_CURRENT_USER\Software\Microsoft\Command Processor\AutoRun

==> For changes to take effect, close and re-open your current shell. <==


(base) C:\Users\dylan>conda init powershell
no change     C:\Users\dylan\Anaconda3\Scripts\conda.exe
no change     C:\Users\dylan\Anaconda3\Scripts\conda-env.exe
no change     C:\Users\dylan\Anaconda3\Scripts\conda-script.py
no change     C:\Users\dylan\Anaconda3\Scripts\conda-env-script.py
no change     C:\Users\dylan\Anaconda3\condabin\conda.bat
no change     C:\Users\dylan\Anaconda3\Library\bin\conda.bat
no change     C:\Users\dylan\Anaconda3\condabin\_conda_activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\rename_tmp.bat
no change     C:\Users\dylan\Anaconda3\condabin\conda_auto_activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\conda_hook.bat
no change     C:\Users\dylan\Anaconda3\Scripts\activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\activate.bat
no change     C:\Users\dylan\Anaconda3\condabin\deactivate.bat
no change     C:\Users\dylan\Anaconda3\Scripts\activate
no change     C:\Users\dylan\Anaconda3\Scripts\deactivate
no change     C:\Users\dylan\Anaconda3\etc\profile.d\conda.sh
no change     C:\Users\dylan\Anaconda3\etc\fish\conf.d\conda.fish
no change     C:\Users\dylan\Anaconda3\shell\condabin\Conda.psm1
no change     C:\Users\dylan\Anaconda3\shell\condabin\conda-hook.ps1
no change     C:\Users\dylan\Anaconda3\Lib\site-packages\xontrib\conda.xsh
no change     C:\Users\dylan\Anaconda3\etc\profile.d\conda.csh
no change     C:\Users\dylan\OneDrive\Documents\WindowsPowerShell\profile.ps1
No action taken.

After closing and reopening the anaconda prompt again for the changes to take effect, I found I had no functional anaconda terminal and just showed this:
[cid:dd7de4c6-b4fc-477e-bb55-c5e40b51c054]

However it did fix all the problem encountered importing modules in a VSCode window, but still hasnt added

C:\\Users\\dylan\\Anaconda3\\Scripts

to the pythonpath - sys.path returns:

['c:\\Users\\dylan\\AppData\\Local\\Temp',
 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles\\vscode_datascience_helpers',
 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles',
 'c:\\Users\\dylan\\.vscode\\extensions\\ms-toolsai.jupyter-2021.6.999406279\\pythonFiles\\lib\\python',
 'C:\\Users\\dylan\\Anaconda3\\python37.zip',
 'C:\\Users\\dylan\\Anaconda3\\DLLs',
 'C:\\Users\\dylan\\Anaconda3\\lib',
 'C:\\Users\\dylan\\Anaconda3',
 '',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\dylan\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\Users\\dylan\\AppData\\Roaming\\Python\\Python37\\site-packages\\IPython\\extensions',
 'C:\\Users\\dylan\\.ipython']


On opening my registry editor I found the conda init commands had added an 'Autorun' registry as shown:
[cid:d4785ece-5e26-46c8-b4cf-dc3782090098]

I found adding the scripts path manually with sys.path.insert(0,'C:\\Users\\dylan\\Anaconda3\\Scripts') has no effect on the import ability of python in a VSCode jupyter notebook.

In summary, it seems that I need to run the conda init commands in order to add some registry allowing me to import modules correctly in jupyter notebook on VSCode. However I would need to delete said registry every time i wanted to use anaconda prompt to open a web browser based jupyter notebook.

This is workable but ultimately a little clunky and inconvenient: Is there some way to fix this?

Best wishes,
Dylan
________________________________

**Comment by rchiodo** [comment]
So I think you're saying running ```conda init cmd.exe``` and ```conda init powershell``` has broken the anaconda prompt? That sounds like you should log a bug on anaconda. That certainly shouldn't break your anaconda prompt. And it's the recommended method for using anaconda (and what we recommend if anaconda isn't working).

**Comment by dylan-behr** [comment]
Thank you Rich, I have logged a bug on anaconda.

Best,
Dylan

________________________________
